src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // check that the body of the forEach does not contain any identifiers from the query
+                foreach (var identifier in queryExpressionProcessingInfo.Identifiers)
+                {
+                    // Identifier from the foreach can already be in scope of the foreach statement.
+                    if (forEachStatement.Identifier.ValueText != identifier.ValueText)
+                    {
+                        var name = identifier.ValueText;
+                        if (_semanticFacts.GenerateUniqueName(
+                            _semanticModel,
+                            location: forEachStatement.Statement,
+                            containerOpt: forEachStatement.Statement,
+                            baseName: name,
+                            usedNames: Enumerable.Empty<string>(), _cancellationToken).ValueText != name)
+                        {
+                            documentUpdateInfo = default;
+                            return false;
+                        }
+                    }
+                }
+
+                // If query does not contains identifier with the same name as declared in the foreach,
+                // declare this identifier in the body.
+                if (!queryExpressionProcessingInfo.ContainsIdentifier(forEachStatement.Identifier))
+                {
+                    documentUpdateInfo = ConvertIfInToForeachWithExtraVariableDeclaration(forEachStatement, queryExpressionProcessingInfo);
+                    return true;
+                }
+                else
+                {
+                    // The last select expression in the query returns this identifier:
+                    // foreach(var thisIdentifier in from ....... select thisIdentifier)
+                    // if thisIdentifier in foreach is var, it is OK
+                    // if a type is specified for thisIdentifier in forEach, check that the type is the same as in the select expression
+                    // foreach(MyType thisIdentifier in from ....... from MyType thisIdentifier ... select thisIdentifier)
+
+                    // The last clause in query stack must be SelectClauseSyntax.
+                    var lastSelectExpression = ((SelectClauseSyntax)queryExpressionProcessingInfo.Stack.Peek()).Expression;
+                    if (lastSelectExpression is IdentifierNameSyntax identifierName)
+                    {
+                        if (forEachStatement.Identifier.ValueText == identifierName.Identifier.ValueText)
+                        {
+                            var forEachStatementTypeSymbol = _semanticModel.GetTypeInfo(forEachStatement.Type, _cancellationToken).Type;
+                            var lastSelectExpressionTypeSymbol = _semanticModel.GetOperation(lastSelectExpression, _cancellationToken).Type;
```

Comment:
>_semanticModel.GetOperation(lastSelectExpression, _cancellationToken).Type [](start = 65, length = 74)

I would use GetTypeInfo(...).ConvertedType API. The way we get type right now possibly doesn't account for implicit conversions.

Reply:
Let's come up with a test.

---
In reply to: [178370753](https://github.com/dotnet/roslyn/pull/25362#discussion_r178370753) [](ancestors = 178370753)

Reply:
Basically we need to check both Type and ConvertedType

---
In reply to: [178370864](https://github.com/dotnet/roslyn/pull/25362#discussion_r178370864) [](ancestors = 178370864,178370753)